Userland/Libraries/LibGfx/ImageFormats/JPEG2000Loader.cpp

```diff
@@ -28,14 +31,142 @@ namespace Gfx {
+    for (size_t i = 0; i < header_box.child_boxes().size(); ++i) {
+        auto const& subbox = header_box.child_boxes()[i];
+        if (subbox->box_type() == ISOBMFF::BoxType::JPEG2000ImageHeaderBox) {
+            if (image_header_box_index.has_value())
+                return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Multiple Image Header boxes");
+            image_header_box_index = i;
+        }
+        if (subbox->box_type() == ISOBMFF::BoxType::JPEG2000ColorSpecificationBox) {
+            if (color_header_box_index.has_value())
+                return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Multiple Color Specification boxes");
+            color_header_box_index = i;
+        }
+    }
+
+    if (!image_header_box_index.has_value())
+        return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Expected Image Header box");
+    if (!color_header_box_index.has_value())
+        return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Expected Color Specification box");
+
+    auto const& image_header_box = static_cast<ISOBMFF::JPEG2000ImageHeaderBox const&>(*header_box.child_boxes()[image_header_box_index.value()]);
+    context.size = { image_header_box.width, image_header_box.height };
+
+    auto const& color_header_box = static_cast<ISOBMFF::JPEG2000ColorSpecificationBox const&>(*header_box.child_boxes()[color_header_box_index.value()]);
+    if (color_header_box.method == 2)
+        context.icc_data = color_header_box.icc_data.bytes();
+
+    return {};
+}
+
 bool JPEG2000ImageDecoderPlugin::sniff(ReadonlyBytes data)
 {
     return data.starts_with(jp2_id_string);
 }
 
-ErrorOr<NonnullOwnPtr<ImageDecoderPlugin>> JPEG2000ImageDecoderPlugin::create(ReadonlyBytes)
+JPEG2000ImageDecoderPlugin::JPEG2000ImageDecoderPlugin()
+{
+    m_context = make<JPEG2000LoadingContext>();
+}
+
+IntSize JPEG2000ImageDecoderPlugin::size()
+{
+    return m_context->size;
+}
+
+ErrorOr<NonnullOwnPtr<ImageDecoderPlugin>> JPEG2000ImageDecoderPlugin::create(ReadonlyBytes data)
+{
+    auto plugin = TRY(adopt_nonnull_own_or_enomem(new (nothrow) JPEG2000ImageDecoderPlugin()));
+    TRY(decode_jpeg2000_header(*plugin->m_context, data));
+    return plugin;
+}
+
+ErrorOr<ImageFrameDescriptor> JPEG2000ImageDecoderPlugin::frame(size_t index, Optional<IntSize>)
+{
+    if (index != 0)
+        return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Invalid frame index");
+
+    if (m_context->state == JPEG2000LoadingContext::State::Error)
+        return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Decoding failed");
+
+    return Error::from_string_literal("JPEG2000ImageDecoderPlugin: Draw the rest of the owl");
+}
+
+ErrorOr<Optional<ReadonlyBytes>> JPEG2000ImageDecoderPlugin::icc_data()
 {
-    return Error::from_string_view("FIXME: Draw the rest of the owl"sv);
+    if (m_context->icc_data.has_value())
+        return *m_context->icc_data;
+    return OptionalNone {};
```

Comment:
Wouldn't this just be `return m_context->icc_data;`?

Reply:
Err! Thanks, done 🤦